### ISSUE TITLE
Allow specifying encodings other than UTF-8 in XML declaration written

### ISSUE THREAD

**Issue by cassiomolin** [FasterXML/jackson-dataformat-xml#315]
The `UTF-8` encoding is hard coded in the [`ToXmlGenerator`](https://fasterxml.github.io/jackson-dataformat-xml/javadoc/2.9/com/fasterxml/jackson/dataformat/xml/ser/ToXmlGenerator.html) source code:

```java
if (Feature.WRITE_XML_1_1.enabledIn(_formatFeatures)) {
    _xmlWriter.writeStartDocument("UTF-8", "1.1");
} else if (Feature.WRITE_XML_DECLARATION.enabledIn(_formatFeatures)) {
    _xmlWriter.writeStartDocument("UTF-8", "1.0");
} else {
    return;
}
```

Once [`ToXmlGenerator`](https://fasterxml.github.io/jackson-dataformat-xml/javadoc/2.9/com/fasterxml/jackson/dataformat/xml/ser/ToXmlGenerator.html) is `final`, there might not be an easy way to have other encodings such as `ISO-8859-1`:

```xml
<?xml version="1.0" encoding="ISO-8859-1"?>
```

See this [question](https://stackoverflow.com/q/52778143/1426227) in Stack Overflow for reference.

**Comment by cowtowncoder** [comment]
Ah. Yes, I see. So although underlying writer may actually use different encoding, xml declaration claims it is UTF-8. That's not good.

**Comment by saimonsez** [comment]
Additionaly, [XmlFactory](https://fasterxml.github.io/jackson-dataformat-xml/javadoc/2.11/com/fasterxml/jackson/dataformat/xml/XmlFactory.html) outputs utf-8 only:

````
protected XMLStreamWriter _createXmlWriter(IOContext ctxt, OutputStream out) throws IOException
    {
     XMLStreamWriter sw;
     try {
         sw = _xmlOutputFactory.createXMLStreamWriter(_decorate(ctxt, out), "UTF-8");
     } catch (Exception e) {
         throw new JsonGenerationException(e.getMessage(), e, null);
     }
     return _initializeXmlWriter(sw);
}

````

**Comment by cowtowncoder** [comment]
@saimonsez That is intentional however (partly since there is no mechanism to pass non-Unicode encodings); caller is expected to create `Writer` for alternate encodings.

However, I hope to introduce a mechanism to allow users to create document "header" (xml declaration and/or DOCTYPE declaration) via `XMLStreamWriter`, which would allow adding encoding in xml declaration.

**Comment by saimonsez** [comment]
I see, thank you for clarification. In my case, the caller is spring-web (AbstractJackson2HttpMessageConverter) without a chance to configure anything, so I am stuck again. Are you by chance involved with springs integration of jackson?

**Comment by saimonsez** [comment]
I just created https://github.com/spring-projects/spring-framework/issues/25076 which is related to this issue (if jackson is used with spring).

**Comment by cowtowncoder** [comment]
I am only involved whenever Spring folks file bugs, but do not know their code base (and they don't use, I think, Jackson JAX-RS provider).
Their involvement would be needed even if new functionality / endpoints were added, for what that is worth.

**Comment by kromit** [comment]
This is still moved for 2 years from `2.10` to `2.11` to `2.12`. So next is`2.13`? 

Any stupid workaround would be great.

**Comment by cowtowncoder** [comment]
@kromit You are absolutely welcome to provide a fix as you seem to need it.

**Comment by kromit** [comment]
@cowtowncoder I've looked into this and I would break significantly more things on the way, than I would fix. :see_no_evil: 
Not sure if my workaround is legit but this is what I am using.

```
private final String DOCTYPE ="<?xml version=\"1.0\" encoding=\"iso-8859-1\"?>\n";
  
Writer writer = new OutputStreamWriter(out, StandardCharsets.ISO_8859_1);
writer.write(DOCTYPE);

XmlMapper xmlMapper = new XmlMapper();
xmlMapper.writeValue(writer, value);
```

**Comment by cowtowncoder** [comment]
Couple of possibly helpful pointers:

* #150 is related; there should be a way to customize writing of DOCTYPE as well as xml declaration
* You could also manually create `XMLStreamWriter`, use `writer.writeStartDocument(...)` to initialize, pass to XML-specific `mapper.writeValue()` method. But if you do so, need to disable `ToXmlGenerator.Feature.WRITE_XML_DECLARATION` (in fact that may already be necessary in your case?)

The idea with #150 (which I really would like to get in 2.12 if I have time) would be to allow registering a writer callback that would write all pre-amble events (xml declaration and/or DOCTYPE) the way caller wants. Conceptually simple just need to think of a way to do that in a way that fits with format-specific handling of Jackson's databind (most API is format-agnostic).

**Comment by pjfanning** [comment]
@cowtowncoder is there any appetite to take this on for v2.16?

If I was to look at this, I'd prefer not to use custom declaration handling and to concentrate on allowing this module to create XMLStreamWriter instances that could be created with other encodings.

In pseudocode, a user might be able to do this
```
XmlMapper mapper = new XmlMapper();
mapper.setDefaultEncoding("Big5");
byte[] bytes = mapper.writeValueAsBytes(myObjectInstance);
```

The `bytes` in the example above would pop out with `Big5` as the encoding in the XML declaration and the chars would be encoded as Big5.
 
If you look at XmlFactory, there is some support for `JsonEncoding` but this enum has very limited values - all Unicode ones. This is why in the pseudocode, the encoding is a String - or it could be a java.nio.charset.Charset. I know that the Java encoding names may not be ideal - as they may not match the encoding names used in other applications but it should be a decent enough place to start.

I have a couple of changes [here](https://github.com/FasterXML/jackson-dataformat-xml/compare/2.16...pjfanning:jackson-dataformat-xml:xml-encoding) but the use of JsonEncoding is not a good place to start. And in practice, users don't typically create their own generators and use higher level methods like writeValueAsString/writeValueAsBytes.

**Comment by cowtowncoder** [comment]
@pjfanning Agreed, `JsonEncoding` is too limited to help. And yes, it'd be great to work on solving this in 2.16. I think your points are valid wrt it being difficult to add simple handler as one also needs to be able to create underlying `XMLStreamWriter` that uses encoding -- although this is not necessarily as problematic for other things one might want to "inject" (like DOCTYPE declaration).

So I am open to alternate takes here; I do not have current specific plans (or time) to tackle this myself, so should not try to dictate solution. But will definitely give feedback :)

**Comment by pjfanning** [comment]
@cowtowncoder the workaround in https://github.com/FasterXML/jackson-dataformat-xml/blob/5063969ec78dcfe076ac3356fdd1eefec7aedcdc/src/test/java/com/fasterxml/jackson/dataformat/xml/incr/IncrementalWritingTest.java  - does this only work with the v2.17 future release?

**Comment by cowtowncoder** [comment]
@pjfanning I think using `XMLStreamWriter` and writing should work with earlier versions, but is not tested. If you wanted to backport tests to verify in 2.16 that'd be fine?

2.17 adds convenience factory methods but fundamentally those are not necessary I think.

This isn't super convenient mechanism of course, but at least makes certain things possible.

**Comment by cowtowncoder** [comment]
Can now be implemented, using https://github.com/FasterXML/jackson-databind/issues/5860 -- need builder-buildable XML-specific `GeneratorInitializer` and that can add desired output.

**Comment by pjfanning** [comment]
ae7e650e510931edad2aacd35af15367df4683ce shows how to do this

**Comment by cowtowncoder** [comment]
@pjfanning True, but I have something bit cleaner in mind; reusable builder-based `GeneratorInitializer` that can be configured with settings to use. This will be necessary with some of initialization issues like binding of namespace mappings. Users can of course write and use their own `GeneratorInitializer`s too.

But it may be necessary to add overrides on `XmlMapper` to ensure initializer is always called.

**Comment by cowtowncoder** [comment]
Can solve similar to #150, using brand new `XmlGeneratorInitializer`; will probably just allow specifying full XML declaration with 3 part:

1. XML version ("1.0" or "1.1"; defaulting to "1.0" if not specified
2. Encoding (defaulting to "UTF-8")
3. Stand-alone-ness: optional, defaults to not output

**Comment by cowtowncoder** [comment]
Mechanism to use:

```java
// can configure either XmlMapper or ObjectWriter with XmlGeneratorInitializer
ObjectWriter w = new XmlMapper.writer().with(new XmlGeneratorInitializer()
    .addXmlDeclaration("1.0", "ISO-8859-1"));
String xml = w.writeValueAsString(new Value());

```
